### PR TITLE
Encapsulating fodf, resample_volume

### PR DESCRIPTION
**Pull request by EmmaRenauld**
Prepared:

- scil_compute_fodf
- scil_resample_volume
- Resample_streamlines (added possibility to resample to a fixed step size instead of a nb of points. Changed in the function (tracking/tools) but the main script is not changed).

Note. This PR will only work after merging Philippe's #96

### CODE REVIEW

**Comment by EmmaRenauld** [comment]
See reconst/fodf ligne 121, I wasn't sure.

File: scilpy/reconst/fodf.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        raise ValueError('FRF file did not contain 4 elements. '
+                         'Invalid or deprecated FRF format')
+    frf = full_frf[0:3]
+    mean_b0_val = full_frf[3]
+
+    # Checking if we will use parallel processing
+    parallel = True
+    if nbr_processes is not None:
+        if nbr_processes <= 0:
+            nbr_processes = None        # C'est ok ça?? parallel reste true si nbr_processes <0? Je laisse comme c'était.
```
Comment (jchoude) [inline scilpy/reconst/fodf.py]: Please write in English.

**Comment by jchoude** [comment]
Please use a more explicit PR name next time... through all the emails we get on PRs, the purpose is not really clear without opening the whole thread.

**Comment by jchoude** [comment]
You need to rebase or merge master here.

File: scilpy/reconst/fodf.py (hunk 1 above)
Comment (EmmaRenauld) [reply scilpy/reconst/fodf.py]: I just forgot to erase it. I actually wanted to ask you. Why do you deal with a negative nbr_processes? Why do you keep parallel=true in that case?

Comment (EmmaRenauld) [reply scilpy/reconst/fodf.py]: @jchoude ?

File: scilpy/image/resample_volume.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+def _interp_code_to_order(interp_code):
+    orders = {'nn': 0, 'lin': 1, 'quad': 2, 'cubic': 3}
+    return orders[interp_code]
+
+
+def resample_volume(img, ref=None, res=None, iso_min=False, interp='lin',
+                    enforce_dimensions=False):
+    """
+    Script to resample a dataset to match the resolution of another
```
Comment (ppoulin91) [inline scilpy/image/resample_volume.py]: Script -> **Function**

File: scilpy/image/resample_volume.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        resample to an isotropic resolution of Y x Y x Y. This method is used
+        only if res is not None. (default: None)
+    iso_min: bool, optional
+        If true, resample the volume to R x R x R with R being the smallest
+        current voxel dimension. If false, this method is not used.
+    interp: str, optional
+        Interpolation mode. 'nn' = nearest neighbour, 'lin' = linear,
+        'quad' = quadratic, 'cubic' = cubic. (Default: linear)
+    enforce_dimensions: bool, optional
+        If True, enforce the reference volume dimension (only if res in not None).
```
Comment (ppoulin91) [inline scilpy/image/resample_volume.py]: only if res **is** not None

File: scilpy/image/resample_volume.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        (Default = False)
+
+    Returns
+    -------
+    resampled_image: nib.Nifti1Image
+        Resampled image.
+    """
+    data = img.get_data()
```
Comment (ppoulin91) [inline scilpy/image/resample_volume.py]: get_fdata()

File: scilpy/image/resample_volume.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    if ref is not None:
+        ref_img = nib.load(ref)
+        new_zooms = ref_img.header.get_zooms()[:3]
+    elif res is not None:
+        new_zooms = [res] * 3
+    elif iso_min:
+        min_zoom = min(original_zooms)
+        new_zooms = (min_zoom, min_zoom, min_zoom)
+    else:
```
Comment (ppoulin91) [inline scilpy/image/resample_volume.py]: ref/ref/iso_min should be mutually exclusive, can you check to make sure there are not multiple options?

File: scilpy/io/utils.py [hunk 6]
```diff
@@ -121,6 +121,16 @@ def add_sh_basis_args(parser, mandatory=False):
                         help=help_msg)
 
 
+def check_sh_basis_choice(sh_basis):
+    """ Check if the passed sh_basis arg to a fct is right.
+    :param sh_basis:
```
Comment (ppoulin91) [inline scilpy/io/utils.py]: Please use numpy-style docstring

File: scilpy/image/resample_volume.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
```
Comment (ppoulin91) [inline scilpy/image/resample_volume.py]: You can remove the shebang

File: scilpy/reconst/fodf.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
+# !/usr/bin/env python
```
Comment (ppoulin91) [inline scilpy/reconst/fodf.py]: Remove shebang

File: scilpy/reconst/fodf.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+# -*- coding: utf-8 -*-
+
+import warnings
```
Comment (ppoulin91) [inline scilpy/reconst/fodf.py]: imports order

File: scilpy/reconst/fodf.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+import logging
+
+from dipy.core.gradients import gradient_table
+from dipy.data import get_sphere
+from dipy.reconst.csdeconv import ConstrainedSphericalDeconvModel
+from dipy.direction.peaks import peaks_from_model
+
+from scilpy.utils.bvec_bval_tools import normalize_bvecs, is_normalized_bvecs
```
Comment (ppoulin91) [inline scilpy/reconst/fodf.py]: Merge all imports from bvec_bval_tools

File: scilpy/reconst/fodf.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    -------
+    peaks_csd: PeaksAndMetrics
+        An object with ``gfa``, ``peak_directions``, ``peak_values``,
+        ``peak_indices``, ``odf``, ``shm_coeffs`` as attributes
+    """
+
+    # Checking data and sh_order
+    check_b0_threshold(force_b0_threshold, bvals.min())
+    if data.shape[-1] < (sh_order + 1) * (sh_order + 2) / 2:
+        warnings.warn(
```
Comment (ppoulin91) [inline scilpy/reconst/fodf.py]: Please use logging

File: scilpy/io/utils.py [hunk 12]
```diff
@@ -121,6 +121,16 @@ def add_sh_basis_args(parser, mandatory=False):
                         help=help_msg)
 
 
+def check_sh_basis_choice(sh_basis):
```
Comment (ppoulin91) [inline scilpy/io/utils.py]: Can we rename it to ""validate_sh_basis_choice". _Check_ does not feel explicit enough for me

File: scripts/scil_resample_volume.py [hunk 13]
```diff
@@ -71,46 +66,22 @@ def main():
...
     img = nib.load(args.input)
-    data = img.get_data()
-    affine = img.get_affine()
-    original_zooms = img.get_header().get_zooms()[:3]
 
+    # Resampling volume
     if args.ref:
-        ref_img = nib.load(args.ref)
-        new_zooms = ref_img.header.get_zooms()[:3]
+        resampled_image = resample_volume(img, ref=args.ref,
```
Comment (ppoulin91) [inline scripts/scil_resample_volume.py]: You could just call it once with all parameters, which should be None if they were not provided

File: scripts/scil_resample_streamlines.py [hunk 14]
```diff
@@ -42,8 +42,8 @@ def main():
     streamlines = list(tractogram_file.streamlines)
 
     new_streamlines = resample_streamlines(streamlines,
-                                           args.nb_pts_per_streamline,
-                                           args.arclength)
+                                           num_points=args.nb_pts_per_streamline,
```
Comment (arnaudbore) [inline scripts/scil_resample_streamlines.py]: forget the new step_size

Comment (arnaudbore) [reply scripts/scil_resample_streamlines.py]: It should also be in mm

Comment (EmmaRenauld) [reply scripts/scil_resample_streamlines.py]: Will be done in PR109.

File: scilpy/image/resample_volume.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+        new_zooms = (min_zoom, min_zoom, min_zoom)
+    else:
+        raise ValueError("You must choose the resampling method. Either with"
+                         "a reference volume, or a chosen isometric resolution,"
+                         "or an isometric resampling to the smallest current "
+                         "voxel dimension!")
+
+    interp_choices = ['nn', 'lin', 'quad', 'cubic']
+    if interp not in interp_choices:
+        raise ValueError("interp must be one of 'nn', 'lin', 'quad', 'cubic'.")
```
Comment (GuillaumeTh) [inline scilpy/image/resample_volume.py]: interp -> --interp

File: scilpy/image/resample_volume.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    data2, affine2 = reslice(data, affine, original_zooms, new_zooms,
+                             _interp_code_to_order(interp))
+
+    logging.debug('Resampled data shape: %s', data2.shape)
+    logging.debug('Resampled data affine: %s', affine2)
+    logging.debug('Resampled data affine setup: %s', nib.aff2axcodes(affine2))
+
+    if enforce_dimensions:
+        if ref is None:
+            raise ValueError('enforce_dimensions can only be used with the ref'
```
Comment (GuillaumeTh) [inline scilpy/image/resample_volume.py]: enforce -> --enforce

File: scilpy/image/resample_volume.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    resampled_image: nib.Nifti1Image
+        Resampled image.
+    """
+    data = img.get_fdata()
+    affine = img.get_affine()
+    original_zooms = img.get_header().get_zooms()[:3]
+
+    if ref is not None:
+        if res is not None or iso_min:
+            raise ValueError('Please only provide one option amongst ref, res '
```
Comment (GuillaumeTh) [inline scilpy/image/resample_volume.py]: --ref, --res or --iso_min

File: scilpy/image/resample_volume.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+                             'or iso_min.')
+        ref_img = nib.load(ref)
+        new_zooms = ref_img.header.get_zooms()[:3]
+    elif res is not None:
+        if iso_min:
+            raise ValueError('Please only provide one option amongst ref, res '
```
Comment (GuillaumeTh) [inline scilpy/image/resample_volume.py]: --ref, --res or --iso_min

File: scilpy/io/utils.py [hunk 19]
```diff
@@ -127,6 +127,24 @@ def add_sh_basis_args(parser, mandatory=False):
...
+    sh_basis: str
+        Either 'descoteaux08' or 'tournier07'
+
+    Raises
+    ------
+    ValueError
+        If sh_basis is not one of 'descoteaux07' or 'tournier07'
+    """
+    if not (sh_basis == 'descoteaux07' or sh_basis == 'tournier07'):
+        raise ValueError("sh_basis should be either 'descoteaux07' or"
```
Comment (GuillaumeTh) [inline scilpy/io/utils.py]: --sh_basis

File: scilpy/reconst/fodf.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+
+    # Checking if we will use parallel processing
+    parallel = True
+    if nbr_processes is not None:
+        if nbr_processes == 0:  # Will use all processed
+            nbr_processes = None
+        elif nbr_processes == 1:
+            parallel = False
+        elif nbr_processes < 0:
+            raise ValueError('nbr_processes should be positive.')
```
Comment (GuillaumeTh) [inline scilpy/reconst/fodf.py]: --nbr_processes

File: scripts/scil_compute_fodf.py [hunk 21]
```diff
@@ -104,69 +99,25 @@ def main():
...
-        raise ValueError('FRF file did not contain 4 elements. '
-                         'Invalid or deprecated FRF format')
-
-    frf = full_frf[0:3]
-    mean_b0_val = full_frf[3]
-
     vol = nib.load(args.input)
-    data = vol.get_data()
-
+    data = vol.get_fdata()
```
Comment (GuillaumeTh) [inline scripts/scil_compute_fodf.py]: The problem with this is that you load in f64. Maybe use dataobj() ? @jchoude @arnaudbore what do you think about that ?

Comment (arnaudbore) [reply scripts/scil_compute_fodf.py]: agree with guillaume we should use dataobj()

**Comment by jhlegarreta** [comment]
@GuillaumeTh the changes you requested seem to have been addressed. Could you please review the PR again? Thanks.

File: scripts/scil_compute_fodf.py [hunk 22]
```diff
@@ -104,69 +99,25 @@ def main():
...
-                                 relative_peak_threshold=.5,
-                                 min_separation_angle=25,
-                                 mask=mask,
-                                 return_sh=True,
-                                 sh_basis_type=args.sh_basis,
-                                 sh_order=args.sh_order,
-                                 normalize_peaks=True,
-                                 parallel=parallel,
-                                 nbr_processes=nbr_processes)
+        mask = np.asanyarray(nib.load(args.mask_wm).dataobj).astype(np.bool)
```
Comment (arnaudbore) [inline scripts/scil_compute_fodf.py]: mask_wm does not exist

Comment (jhlegarreta) [reply scripts/scil_compute_fodf.py]: I did not check all options/conditions, and I did not specifically provide a mask, which explains why my checks were passing. I am sorry @arnaudbore .

**Pull request merged**